common/models/user.js

```diff
@@ -468,130 +468,157 @@ module.exports = function(User) {
-    assert(typeof tokenGenerator === 'function', 'generateVerificationToken must be a function');
+    // assert the verifyOptions params that might have been badly defined
+    assertVerifyOptions(verifyOptions);
 
-    tokenGenerator(user, function(err, token) {
-      if (err) { return fn(err); }
+    // argument "options" is passed depending on verifyOptions.generateVerificationToken function requirements
+    var tokenGenerator = verifyOptions.generateVerificationToken;
+    if (tokenGenerator.length == 3) {
+      tokenGenerator(user, options, addTokenToUserAndSave);
+    } else {
+      tokenGenerator(user, addTokenToUserAndSave);
+    }
 
+    function addTokenToUserAndSave(err, token) {
+      if (err) return cb(err);
       user.verificationToken = token;
-      user.save(function(err) {
-        if (err) {
-          fn(err);
-        } else {
-          sendEmail(user);
-        }
+      user.save(options, function(err) {
+        if (err) return cb(err);
+        sendEmail(user);
       });
-    });
+    }
 
     // TODO - support more verification types
     function sendEmail(user) {
-      options.verifyHref += '&token=' + user.verificationToken;
+      verifyOptions.verifyHref += '&token=' + user.verificationToken;
+      verifyOptions.verificationToken = user.verificationToken;
 
-      options.verificationToken = user.verificationToken;
+      verifyOptions.text = verifyOptions.text || g.f('Please verify your email by opening ' +
+        'this link in a web browser:\n\t%s', verifyOptions.verifyHref);
 
-      options.text = options.text || g.f('Please verify your email by opening ' +
-        'this link in a web browser:\n\t%s', options.verifyHref);
+      verifyOptions.text = verifyOptions.text.replace(/\{href\}/g, verifyOptions.verifyHref);
 
-      options.text = options.text.replace(/\{href\}/g, options.verifyHref);
+      // argument "options" is passed depending on templateFn function requirements
+      var templateFn = verifyOptions.templateFn;
+      if (templateFn.length == 3) {
+        templateFn(verifyOptions, options, setHtmlContentAndSend);
+      } else {
+        templateFn(verifyOptions, setHtmlContentAndSend);
+      }
 
-      options.to = options.to || user.email;
+      function setHtmlContentAndSend(err, html) {
+        if (err) return cb(err);
 
-      options.subject = options.subject || g.f('Thanks for Registering');
+        verifyOptions.html = html;
 
-      options.headers = options.headers || {};
+        // Remove verifyOptions.template to prevent rejection by certain
+        // nodemailer transport plugins.
+        delete verifyOptions.template;
 
-      options.templateFn(options, function(err, html) {
-        if (err) {
-          fn(err);
+        // argument "options" is passed depending on Email.send function requirements
+        var Email = verifyOptions.mailer;
+        if (Email.send.length == 3) {
+          Email.send(verifyOptions, options, handleAfterSend);
```

Comment:
Email.send does not support options yet - see https://github.com/strongloop/loopback/blob/95240598b3d7d3b365eaa184f53a0ae26a6aad14/lib/connectors/mail.js#L142

It would be nice to eventually fix Email's API and add support for options too. Let's leave that out of scope of this pull request though.